ledger/participant-integration-api/src/main/scala/platform/indexer/MeteringAggregator.scala

```diff
@@ -0,0 +1,180 @@
+import java.util.{Timer, TimerTask}
+import scala.concurrent.duration._
+import scala.concurrent.{Await, ExecutionContext, Future}
+import scala.util.{Failure, Success}
+
+object MeteringAggregator {
+
+  class Owner(
+      meteringStore: MeteringStorageWriteBackend,
+      parameterStore: ParameterStorageBackend,
+      metrics: Metrics,
+      period: FiniteDuration = 6.minutes,
+  ) {
+
+    private[platform] def apply(
+        dispatcher: DbDispatcher
+    )(implicit loggingContext: LoggingContext): ResourceOwner[Unit] = {
+      val aggregator = new MeteringAggregator(meteringStore, parameterStore, metrics)
+      Await.ready(aggregator.initialize(dispatcher), 60.seconds)
+      ResourceOwner.forTimer(() => new Timer()).map { timer =>
+        timer.scheduleAtFixedRate(
+          new TimerTask {
+            override def run(): Unit = {
+              Await.ready(aggregator.run(dispatcher), 60.minutes)
+              ()
+            }
+          },
+          period.toMillis,
+          period.toMillis,
+        )
+      }
+    }
+  }
+
+  private def toTimestamp(dateTime: OffsetDateTime): Timestamp =
+    Timestamp.assertFromInstant(dateTime.toInstant)
+  private def toOffsetDateTime(timestamp: Timestamp): OffsetDateTime =
+    OffsetDateTime.ofInstant(timestamp.toInstant, ZoneOffset.UTC)
+
+}
+
+class MeteringAggregator(
+    meteringStore: MeteringStorageWriteBackend,
+    parameterStore: ParameterStorageBackend,
+    metrics: Metrics,
+    clock: () => Timestamp = () => Timestamp.now(),
+)(implicit loggingContext: LoggingContext) {
+
+  private implicit val ec: ExecutionContext = ExecutionContext.global
+
+  private val logger = ContextualizedLogger.get(getClass)
+
+  private[platform] def initialize(sqlExecutor: SqlExecutor): Future[Unit] = {
+    val initTimestamp = toOffsetDateTime(clock()).truncatedTo(ChronoUnit.HOURS).minusHours(1)
+    val initLedgerMeteringEnd = LedgerMeteringEnd(Offset.beforeBegin, toTimestamp(initTimestamp))
+    sqlExecutor.executeSql(metrics.daml.index.db.initializeMeteringAggregator) {
+      parameterStore.initializeLedgerMeteringEnd(initLedgerMeteringEnd)
+    }
+  }
+
+  private[platform] def run(sqlExecutor: SqlExecutor): Future[Unit] = {
+
+    val future = sqlExecutor.executeSql(metrics.daml.index.db.meteringAggregator) { conn =>
+      val nowUtcTime = toOffsetDateTime(clock())
+      val lastLedgerMeteringEnd = getLedgerMeteringEnd(conn)
+      val startUtcTime: OffsetDateTime = toOffsetDateTime(lastLedgerMeteringEnd.timestamp)
+      val endUtcTime = startUtcTime.plusHours(1)
+
+      if (nowUtcTime.isAfter(endUtcTime)) {
+
+        val toEndTime = toTimestamp(endUtcTime)
+        val ingestedLedgerEnd = parameterStore.ledgerEnd(conn).lastOffset
+        val maybeMaxOffset =
+          meteringStore.transactionMeteringMaxOffset(lastLedgerMeteringEnd.offset, toEndTime)(conn)
+
+        val (
+          periodIngested, // This is true if the time period is closed fully ingested
```

Comment:
I am not sure about this.

In the akka-stream pipeline the timestamps are assigned in an async stage at input ingestion, which is followed by two buffers and another asynchronous stage, before getting to an asynchronous ingestion stage, where it is getting put into the database. IMHO it is very well possible (although not likely I agree) that by the time, we here decide that we have all metering_transactions, we still have some getting ingested, which have time earlier than toEndTime.

I am not sure this is a problem though, because we anyway we will pick up these renegade entries with next aggregation (although it would be a small lie stating that it was in the recorded period, but I think it is ok)

On any rate: I do not understand having here special treatments with considering ledgerEnd.
Can it be that just switching the transactionMeteringMaxOffset respecting the ledgerEnd as we usually do would make this code clearer, with same guarantees?

On aggregating "renegate" entries in the next aggregation: if that would be a problem, I would probably just add some reasonable time to the condition at 89

Reply:
Other parts of the code base are initialized with a LedgerEndCache.  As this processing initiated as part of the ingestion we do not have that cache (and there is only one!).  Hence this approach.

Reply:
@mziolekda agrees not to make any changes to this for this PR.